### PR TITLE
WIP: Add presenters

### PR DESCRIPTION
**Pull request by timurvafin**
Introduce presenters in to rails-base.

* PORO based
* `app/presenters` directory to store presenters
* `BasePresenter` class to present single object
*  `BaseCollectionPresenter` class to present collection
* `decent_decoration` gem to expose AR instance decorated for view and clean inside controller with that same name
* `current_user` presented with `UserPresenter`

Here is controller example:

```ruby
class UnitsController < ApplicationController
  before_action :authenticate_user!

  responders :collection
  respond_to :html

  expose_decorated(:units, decorator: UnitsPresenter) { Unit.list(params[:page]) }
  expose_decorated(:unit, attributes: :unit_params, decorator: UnitPresenter)

  def index
  end

  def new
  end

  def create
    unit.save
    respond_with(unit)
  end

  def edit
  end

  def update
    unit.save
    respond_with(unit)
  end

  def destroy
    unit.destroy
    respond_with(unit)
  end

  private

  def unit_params
    params.require(:unit).permit(:inv_id, :unit_type, :name, :description, :location_id, :user_id)
  end
end
```

Form example.
We need still use `to_model` inside `form_for` while [pull request](https://github.com/plataformatec/simple_form/pull/1256) to fix that not merged.

```ruby
.row
  .columns
    h2 = title

.row
  .medium-6.columns
    = simple_form_for(unit.to_model) do |f|
      .form-inputs
        = f.input :inv_id, required: true, autofocus: true
        = f.input :unit_type
        = f.association :user, label_method: :full_name
        = f.association :location
        = f.input :name, required: true
        = f.input :description, as: :text, required: false
      .form-actions
        = f.button :submit
```

List view example:
```ruby
.row
  .columns
    h2 = title

.row
  .medium-6.columns
    table
      thead
        tr
          th Type
          th Location
          th User
          th Name
          th Tools
      tbody
        = render(units_list)

    = paginate(units_list)
    = link_to('New unit', new_unit_path, class: 'button')
```

### CODE REVIEW

**Comment by timurvafin** [comment]
@vast @VladimirMikhailov @dmitrytrager check this out

File: app/presenters/model_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class ModelPresenter
+  # ActiveModel compability
```
Comment (dmitrytrager) [inline app/presenters/model_presenter.rb]: compatibility)

**Comment by dmitrytrager** [comment]
:ok:

File: app/presenters/users_presenter.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class UsersPresenter < CollectionPresenter
+  self.presenter = UserPresenter
+
+  def avarage_something
```
Comment (vast) [inline app/presenters/users_presenter.rb]: avErage. Shouldn't we have this one in `examples` branch?

**Comment by timurvafin** [comment]
@vast `UsersPresenter` removed

File: app/presenters/user_presenter.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class UserPresenter < ModelPresenter
+  # Devise compability
```
Comment (vast) [inline app/presenters/user_presenter.rb]: same here: compaTIbility

**Comment by vast** [comment]
:ok: except minor typo

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -5,6 +5,16 @@ class ApplicationController < ActionController::Base
   responders :flash
   rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
 
+  helper do
```
Comment (timurvafin) [inline app/controllers/application_controller.rb]: @vast this piece of code looks like :hankey: Any thought how to make it better?

Comment (vast) [inline app/controllers/application_controller.rb]:
```ruby
def current_user
  UserPresenter.new(warden.authenticate(scope: :user))
end
```

Comment (timurvafin) [inline app/controllers/application_controller.rb]: @vast we will have `current_user` wrapped in the controller then

**Comment by timurvafin** [comment]
@VladimirMikhailov @vast @dmitrytrager checkout updated presenter

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (vast) [inline app/controllers/application_controller.rb]: @timurvafin, yep, missed that.

Another option is to have `CurrentUserHelper`:

```ruby
class ApplicationController
  helper :current_user
end
```

```ruby
module CurrentUserHelper
  def current_user
    UserPresenter.new(warden.authenticate(scope: :user))
  end
end
```

Comment (RamilGilmanov) [inline app/controllers/application_controller.rb]: @timurvafin i think solution with a helper module is a better approach. The controller would remain smaller and clearer.

File: app/presenters/collection_presenter.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class CollectionPresenter
+  include Enumerable
+
+  attr_reader :collection
+  private :collection
+
+  cattr_accessor :presenter
```
Comment (RamilGilmanov) [inline app/presenters/collection_presenter.rb]: @timurvafin I didn't quite get it. Could you please clarify, what is the reason for using cattr_accessor for item presenter?

Comment (timurvafin) [inline app/presenters/collection_presenter.rb]: @RamilGilmanov `CollectionPresenter` should know how what presenter use for it's items.
Basically that was wrong solution, checkout updated version

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (timurvafin) [inline app/controllers/application_controller.rb]: @vast @RamilGilmanov any thoughts about that https://gist.github.com/timurvafin/816dd58467cbc955feec

Comment (vast) [inline app/controllers/application_controller.rb]: @timurvafin, looks good

Comment (RamilGilmanov) [inline app/controllers/application_controller.rb]: @timurvafin that's great!

File: app/presenters/collection_presenter.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class CollectionPresenter
```
Comment (RamilGilmanov) [inline app/presenters/collection_presenter.rb]: Another thing I am concerned about is that the name for a base item presenter class called BasePresenter, whilst class for collection is CollectionPresenter. Feels a bit confusing. What if they were ItemPresenter and CollectionPresenter, or BasePresenter and BaseCollectionPresenter. What do you think? @timurvafin @vast

Comment (vast) [inline app/presenters/collection_presenter.rb]: @RamilGilmanov, +1 on naming them `BasePresenter` and `BaseCollectionPresenter`. In that case we distinguish collection presenter and emphasize that these are base presenters used as parent for all application presenters.

**Comment by timurvafin** [comment]
Closed in favor of https://github.com/fs/rails-base/pull/373

**Pull request closed without merging**